src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -157,6 +182,50 @@ private static IEnumerable<ErrorDiagnostic> GetArrayAssignmentDiagnostics(ITypeM
                     skipTypeErrors: true)); 
         }
 
+        private static IEnumerable<ErrorDiagnostic> GetDiscriminatedObjectAssignmentDiagnostics(ITypeManager typeManager, ObjectSyntax expression, DiscriminatedObjectType targetType, bool skipConstantCheck)
+        {
+            // if we have parse errors, there's no point to check assignability
+            // we should not return the parse errors however because they will get double collected
+            if (expression.HasParseErrors())
+            {
+                yield break;
+            }
+
+            var propertyMap = expression.ToPropertyDictionary();
+
+            if (!propertyMap.TryGetValue(targetType.DiscriminatorKey, out var discriminatorProperty))
+            {
+                // object doesn't contain the discriminator field
+                yield return DiagnosticBuilder.ForPosition(expression).MissingDiscriminator(targetType.DiscriminatorKey, targetType.UnionMembersByKey.Keys);
+                yield break;
+            }
+
+            // At some point in the future we may want to relax the expectation of a string literal key, and allow a generic string.
```

Comment:
>string [](start = 85, length = 6)

Will need to support integers too at some point.

Reply:
Initial goal is for feature parity with swagger discriminators. Do you happen to know of any non-string examples in the specs repo?

Reply:
Not off the type of my head. Or maybe I'm remembering wrong and we don't have to ever do that 🙂.